### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by GranadaBrayan**
Integration of the devise gem in the Blog app, for user authentication; validates the creation, view, and deletion of articles and comments.

### CODE REVIEW

**Comment by estebancordoba** [comment]
could you please do "rebase" with the developer branch?

File: app/controllers/articles_controller.rb [hunk 1]
```diff
@@ -1,6 +1,5 @@
 class ArticlesController < ApplicationController
-  http_basic_authenticate_with name: 'user', password: 'secret', except: %i[index show]
-
+  before_action :authenticate_user!
```
Comment (estebancordoba) [inline app/controllers/articles_controller.rb:2]: What does the `authenticate_user` method do?

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
```
Comment (estebancordoba) [inline app/models/user.rb:4]: is it necessary to include these methods?

File: app/views/articles/index.html.erb [hunk 3]
```diff
@@ -7,21 +7,25 @@
...
       <% @articles.each do |article| %>
-        <div class="card mb-4">
-          <div class="card-body">
-            <div class="row justify-content-between">
-              <div class="col-4">
-                <div class="small text-muted"><%= article.created_at %></div>
-              </div>
-              <div class="col-4">
-                Status: <%= article.status %>
+        <%if article.user == current_user || article.status == 'public'%>
```
Comment (estebancordoba) [inline app/views/articles/index.html.erb:10]: Could we have a method in the controller that does this?

File: app/views/articles/show.html.erb [hunk 4]
```diff
@@ -13,14 +13,18 @@
                   class:"btn btn-outline-primary btn-lg" %></li>
     </div>
     <div class="col-4"><br>
-      <li><%= link_to "Edit", edit_article_path(@article),
+      <%if @article.user == current_user%>
```
Comment (estebancordoba) [inline app/views/articles/show.html.erb:16]: Could we have a method in the controller that does this?

File: app/views/comments/_comment.html.erb [hunk 5]
```diff
@@ -2,33 +2,37 @@
...
-    <blockquote class="blockquote mb-0">
-      <p><%= comment.commenter %></p>
-      <footer class="blockquote-footer"><%= comment.body %></footer>
-      <div class="small text-muted">
-        <div class="row">
-          <div class="col-12 col-sm-6 col-md-8"></div>
-          <div class="col-6 col-md-4">
-            <div class="card-footer text-muted">
-              <%= comment.created_at %>
+  <%if comment.user == current_user || comment.status == 'public'%>
```
Comment (estebancordoba) [inline app/views/comments/_comment.html.erb:12]: same

File: app/views/comments/_comment.html.erb [hunk 6]
```diff
@@ -2,33 +2,37 @@
...
-      </div>
-    </blockquote>
-  </div>
+      </blockquote>
+    </div>
+  <%end%>
 </div>
 <p>
-  <%= link_to 'Delete Comment', [comment.article, comment],
+  <%if comment.user == current_user%>
```
Comment (estebancordoba) [inline app/views/comments/_comment.html.erb:32]: same

File: app/views/devise/passwords/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Send me reset password instructions" %>
```
Comment (estebancordoba) [inline app/views/devise/passwords/new.html.erb:12]: How would the instructions be sent? via email?

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -18,8 +18,12 @@
                 aria-label="Toggle navigation"><span class="navbar-toggler-icon"></span></button>
         <div class="collapse navbar-collapse" id="navbarSupportedContent">
           <ul class="navbar-nav ms-auto mb-2 mb-lg-0">
+            <%if user_signed_in? %>
```
Comment (estebancordoba) [inline app/views/layouts/application.html.erb:21]: having the methods in the controller is more practical, in the comments above you could apply something similar

File: config/initializers/devise.rb [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+
+  # ==> Controller configuration
+  # Configure the parent class to the devise controllers.
+  # config.parent_controller = 'DeviseController'
+
+  # ==> Mailer Configuration
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (estebancordoba) [inline config/initializers/devise.rb:27]: need to configure this?

File: app/controllers/articles_controller.rb (hunk 1 above)
Comment (GranadaBrayan) [reply app/controllers/articles_controller.rb:2]: The `authenticate_user` method prevents the code flow from continuing until a login has been performed.

File: app/models/user.rb (hunk 2 above)
Comment (GranadaBrayan) [reply app/models/user.rb:4]: These are the basic methods for managing user registration and login.

File: app/views/articles/index.html.erb (hunk 3 above)
Comment (GranadaBrayan) [reply app/views/articles/index.html.erb:10]: Yes, to make the code view more aesthetically pleasing.

File: app/views/articles/show.html.erb (hunk 4 above)
Comment (GranadaBrayan) [reply app/views/articles/show.html.erb:16]: Yes, to make the code view more aesthetically pleasing.

File: app/views/comments/_comment.html.erb (hunk 5 above)
Comment (GranadaBrayan) [reply app/views/comments/_comment.html.erb:12]: Yes sir

File: app/views/comments/_comment.html.erb (hunk 6 above)
Comment (GranadaBrayan) [reply app/views/comments/_comment.html.erb:32]: Yes sir

File: app/views/devise/passwords/new.html.erb (hunk 7 above)
Comment (GranadaBrayan) [reply app/views/devise/passwords/new.html.erb:12]: Yes, I would have to configure the default URL options for Devise mail in each environment.

File: app/views/layouts/application.html.erb (hunk 8 above)
Comment (GranadaBrayan) [reply app/views/layouts/application.html.erb:21]: Yes sir

File: config/initializers/devise.rb (hunk 9 above)
Comment (GranadaBrayan) [reply config/initializers/devise.rb:27]: Yes sir, to send mails through Devise you have to configure it, a guide that could help would be [this one](https://www.bogotobogo.com/RubyOnRails/RubyOnRails_Devise_Authentication_Sending_Confirmation_Email.php)

**Pull request merged**